### PR TITLE
v.1.2 proposal

### PR DESCRIPTION
**Pull request by BridgeAR**
I added a build sanity check and added `MIGRATE [...] KEYS key1, ...` support.
Plus some small cleanup.

### CODE REVIEW

File: tools/build.js [hunk 1]
```diff
@@ -2,20 +2,29 @@ var fs = require('fs');
 var path = require('path');
 var stringify = require('json-stable-stringify');
 var commandPath = path.join(__dirname, '..', 'commands.json');
+var redisCommands = require('../');
 
 var Redis = require('ioredis');
 var redis = new Redis(process.env.REDIS_URI);
 
-redis.command(function (err, res) {
+return redis.command().then(function (res) {
```
Comment (luin) [inline tools/build.js]: Why `return` is needed here?

Comment (BridgeAR) [inline tools/build.js]: It is not. It's only my habit of adding a return in front of promises, if not explicitly unwanted.

File: index.js [hunk 2]
```diff
@@ -114,20 +116,28 @@ exports.getKeyIndexes = function (commandName, args, options) {
       }
     }
     break;
-  case 'zunionstore':
-  case 'zinterstore':
-    keys.push(0);
-    keyStop = Number(args[1]) + 2;
-    for (i = 2; i < keyStop; i++) {
+  case 'migrate':
+    for (i = 3; i < args.length - 1; i++) {
```
Comment (luin) [inline index.js]: According to the doc, the third argument could also be a key: `MIGRATE host port key|"" destination-db timeout [COPY] [REPLACE] [KEYS key [key ...]]`. Shouldn't this be:

```javascript
if (args[2] === '' || args[2] === null || typeof args[2] === 'undefined') {
  for (i = 5; i < args.length - 1; i++) {
    if (args[i].toUpperCase() === 'KEYS') {
      for (var j = i + 1; j < args.length; j++) {
        keys.push(j);
      }
      break;
    }
  }
} else {
  keys.push(2);
}
```

Comment (BridgeAR) [inline index.js]: Definitely, I misread some part in the docs.
But `null` and `undefined` would be replaced with a string in node_redis. ioredis converts them to an empty string, right?

Comment (luin) [inline index.js]: Yes, I just removed them...simply checking `''` should do the trick.

File: tools/build.js [hunk 3]
```diff
@@ -2,20 +2,29 @@ var fs = require('fs');
...
-    throw err;
-  }
+  // Find all special handled cases
+  var movableKeys = String(redisCommands.getKeyIndexes).match(/case '[a-z-]+':/g).map(function (entry) {
+    return entry.replace(/^case \'|\':$/g, '');
+  });
 
   var commands = res.reduce(function (prev, current) {
+    if (current[2].indexOf('movablekeys') !== -1) {
+      movableKeys.splice(movableKeys.indexOf(current[0]), 1);
```
Comment (BridgeAR) [inline tools/build.js]: Note to myself: a check, if `current[0]` exists is missing. Otherwise this might result in false positives.

**Comment by BridgeAR** [comment]
Comments incorporated

**Comment by luin** [comment]
Cool! And we'd better add some tests for `migrate` command:

```javascript
expect(index('migrate', ['127.0.0.1', 6379, 'foo', 0, 0, 'COPY'])).to.eql([2]);
expect(index('migrate', ['127.0.0.1', 6379, '', 0, 0, 'REPLACE', 'KEYS', 'foo', 'bar'])).to.eql([7, 8]);
expect(index('migrate', ['127.0.0.1', 6379, '', 0, 0, 'KEYS', 'foo', 'bar'])).to.eql([6, 7]);
```

**Comment by BridgeAR** [comment]
Tests added including a posttest for 100% coverage by adding istanbul.

While being on it I also updated ioredis and found that `.getKeyIndexes()` default had a check that was redundant and the other one could only be met by invalid argument length for that command. As this would result in an error anyway, I decided to remove that check.

**Comment by luin** [comment]
image

**Comment by BridgeAR** [comment]
Thanks for the review :smiley:

**Pull request merged**